### PR TITLE
Form validation

### PR DESCRIPTION
**Pull request by JanaHaeusler**
Hi there! My form now validates the inputs before submitting them. I am happy about any comments and ideas. Thank you! :)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,6 +9,7 @@
     "jest-styled-components": "^7.0.3",
     "react": "^17.0.1",
     "react-dom": "^17.0.1",
+    "react-hook-form": "^6.11.4",
```
Comment (marwinburesch) [inline package.json]: So you've installed this, but didn't use it?

File: src/components/SaveGameForm.js [hunk 2]
```diff
@@ -87,22 +89,70 @@ export default function SaveGameForm({onSubmit}) {
...
+                winner: '',
+                shots:'',
+            })
+            setErrorMessage('')
+        } else {
+           setErrorMessage(validation.message)
+        }
+    }
+
+    function validateForm(formInput) {
```
Comment (marwinburesch) [inline src/components/SaveGameForm.js]: I like that you made your own form validation, but I have some extra thoughts to share on that.

I think you should consider validating the user input on a different event, maybe switching input fields or `onChange`, or maybe `onChange` but delayed.
Also, it might be annoying, but the date could use some validation as well, just to ensure your users are not tampering with the data and potentially wrecking stuff. You could do that separately or in another story if you think it is not vital right now, but I would encourage you simply because of how dates are handled here.

File: src/components/SaveGameForm.js [hunk 3]
```diff
@@ -87,22 +89,70 @@ export default function SaveGameForm({onSubmit}) {
...
+           }
+       } else {
+           return filledInputs
+       }
+    }
+
+    function validateNotEmpty({location, date, players, winner, shots}) {
+       const allInputsFilled = location.trim().length > 0 && date.trim().length > 0 && players.trim().length > 0 && winner.trim().length > 0 && shots.trim() > 0
+       if (allInputsFilled) {
+            return {result: true}
```
Comment (marwinburesch) [inline src/components/SaveGameForm.js]: I think `result` is incorrect here, why not `isValid`? Even if the validation returns that the input is not valid, it is still a result :)

File: src/components/SaveGameForm.js [hunk 4]
```diff
@@ -87,22 +89,70 @@ export default function SaveGameForm({onSubmit}) {
...
+
+       if (filledInputs.result) {
+           if (validNumber.result) {
+                return {result: true}
+           } else {
+               return validNumber
+           }
+       } else {
+           return filledInputs
+       }
```
Comment (marwinburesch) [inline src/components/SaveGameForm.js]: Can't you make this a bit more compact with logical operators?

**Review by marwinburesch — CHANGES_REQUESTED** [review]
Very nice, keep it up :)
🐜

File: package.json (hunk 1 above)
Comment (JanaHaeusler) [reply package.json]: I used it, but decided to go without and forgot about the package.json. Thanks! :)

File: src/components/SaveGameForm.js (hunk 3 above)
Comment (JanaHaeusler) [reply src/components/SaveGameForm.js]: Ah sure, because it is boolean. Sounds good. :)

File: src/components/SaveGameForm.js (hunk 4 above)
Comment (JanaHaeusler) [reply src/components/SaveGameForm.js]: One idea could be to rewrite it into two terneries. But I am not sure, if this is easier to follow, when two terneries are nested within each other? 
``const errorMessage = filledInputs.isValid ? (validNumber.isValid ? {isValid: true} : validNumber) : filledInputs``

File: src/components/SaveGameForm.js (hunk 2 above)
Comment (JanaHaeusler) [reply src/components/SaveGameForm.js]: Regarding the date validation: Are you referring to RegEx in order to check the date? Right now it is really just checking if it is empty or not. I'll look for a better validation. :)

File: src/components/SaveGameForm.js [hunk 5]
```diff
@@ -1,150 +1,297 @@
...
+        function handleChange(event) {
+            setFormInput({
+                ...formInputs,
+                [event.target.name]: event.target.value
+            })
+        }
+       
+        function handleSubmit(event) {
+            event.preventDefault()
+            trimInputs(formInputs)
```
Comment (monroho) [inline src/components/SaveGameForm.js:118]: Nice that you made an extra function for that!

File: src/components/SaveGameForm.js [hunk 6]
```diff
@@ -1,150 +1,297 @@
...
+                        Location
+                    </label>
+                    <input 
+                        type="text" 
+                        name="location"
+                        id="location"
+                        placeholder="Type location ..."
+                        value={formInputs.location}
+                        onChange={handleChange}
+                        onBlur={() => validateLocation(formInputs.location)}
```
Comment (monroho) [inline src/components/SaveGameForm.js:49]: Good UX to validate just on Blur and not on Change!

File: src/hooks/useGameData.js [hunk 7]
```diff
@@ -1,13 +1,13 @@
...
+import SaveLocally from '../lib/SaveLocally'
+import Loadlocally from '../lib/LoadLocally'
 
 export default function useGameData() {
 
-    const [savedGameProfiles, setSavedGameProfiles] = useState(loadlocally('gameProfiles') ?? [])
+    const [savedGameProfiles, setSavedGameProfiles] = useState(Loadlocally('gameProfiles') ?? [])
     
-    saveLocally('gameProfiles', savedGameProfiles)
+    SaveLocally('gameProfiles', savedGameProfiles)
```
Comment (monroho) [inline src/hooks/useGameData.js]: Any Reason you started the "localStorage functions" with a Capital Letter? Just Curios..

**Review by monroho — COMMENTED** [review]
Nice work Jana!

File: src/components/Button.js [hunk 8]
```diff
@@ -3,11 +3,12 @@ import PropTypes from 'prop-types'
 
 Button.propTypes = {
     children: PropTypes.string.isRequired,
+    disabled: PropTypes.bool.isRequired,
   }
 
-export default function Button({children}) {
+export default function Button({children, disabled}) {
```
Comment (badabam) [inline src/components/Button.js]: You don't need this outer function. Just directly
``` 
export default styled.button``
```

File: src/components/SaveGameForm.test.js [hunk 9]
```diff
@@ -30,6 +38,85 @@ describe('SaveGameForm', () => {
...
+    const { getByLabelText, getByRole} = render(<SaveGameForm onSubmit={onSubmitMock} />)
+    
+    user.type(getByLabelText('Location'), '    ') 
+    user.type(getByLabelText('Date'), '') 
+    user.type(getByLabelText('Player(s)'), ' ') 
+    user.type(getByLabelText('Winner(s)'), '  ') 
+    user.type(getByLabelText('Total Shots Winner(s)'), '5') 
+
+    const button = getByRole('button')
+
```
Comment (badabam) [inline src/components/SaveGameForm.test.js]: Please remove extra empty lines

File: src/components/SaveGameForm.test.js [hunk 10]
```diff
@@ -30,6 +38,85 @@ describe('SaveGameForm', () => {
...
+  it('checks if shots input is a number less than 127', () => {
+    const onSubmitMock = jest.fn()
+    const { getByLabelText, getByRole} = render(<SaveGameForm onSubmit={onSubmitMock} />)
+    
+    user.type(getByLabelText('Location'), 'Horner Racecourse') 
+    user.type(getByLabelText('Date'), '2020-11-21') 
+    user.type(getByLabelText('Player(s)'), 'John, Jane') 
+    user.type(getByLabelText('Winner(s)'), 'Jane') 
+    user.type(getByLabelText('Total Shots Winner(s)'), '140')
```
Comment (badabam) [inline src/components/SaveGameForm.test.js]: You should test the edge cases 126 and 127

File: src/components/SaveGameForm.test.js [hunk 11]
```diff
@@ -30,6 +38,85 @@ describe('SaveGameForm', () => {
...
+    user.type(getByLabelText('Date'), '') 
+    user.type(getByLabelText('Player(s)'), '') 
+    user.type(getByLabelText('Winner(s)'), '   ') 
+    user.type(getByLabelText('Total Shots Winner(s)'), '') 
+
+    const button = getByRole('button')
+
+    expect(button).toBeDisabled()
+
+    user.click(getByRole('button'))
```
Comment (badabam) [inline src/components/SaveGameForm.test.js]: You could reuse the const button and not search it again

File: src/hooks/useGameData.js [hunk 12]
```diff
@@ -1,13 +1,13 @@
 import {useState} from 'react'
 import { v4 as uuid } from 'uuid'
-import saveLocally from '../lib/saveLocally'
-import loadlocally from '../lib/loadLocally'
+import SaveLocally from '../lib/SaveLocally'
```
Comment (badabam) [inline src/hooks/useGameData.js]: Should both be lowercase as they are no components.

Comment (badabam) [reply src/hooks/useGameData.js]: Filename and function name should be lowercase.

**Review by badabam — COMMENTED** [review]
Found some small things you could improve

File: src/hooks/useGameData.js (hunk 7 above)
Comment (JanaHaeusler) [reply src/hooks/useGameData.js]: It was a misunderstanding for my part. :D Is changed. Thanks for the hint!

File: src/hooks/useGameData.js (hunk 12 above)
Comment (JanaHaeusler) [reply src/hooks/useGameData.js]: That was a misunderstanding for my part and makes definitively more sense. :) I found the "l" what you meant. Thanks!

**Review by FabianAlleblas — COMMENTED** [review]
Good job 👍 , nothing to comment.
Maybe you could make one localStorage file for both functions, but that's just my personal liking...

File: src/components/SaveGameForm.test.js [hunk 13]
```diff
@@ -4,32 +4,122 @@ import SaveGameForm from './SaveGameForm'
...
+    const { getByLabelText, getByRole} = render(<SaveGameForm onSubmit={onSubmitMock} />)
+    user.type(getByLabelText('Location'), '   ') 
+    user.type(getByLabelText('Date'), '') 
+    user.type(getByLabelText('Player(s)'), '') 
+    user.type(getByLabelText('Winner(s)'), '   ') 
+    user.type(getByLabelText('Total Shots Winner(s)'), '') 
+    const button = getByRole('button')
+    expect(button).toBeDisabled()
+    user.click(button)
+    expect(onSubmitMock).not.toHaveBeenCalled()
```
Comment (ninapeters) [inline src/components/SaveGameForm.test.js:123]: 👏

**Review by ninapeters — COMMENTED** [review]
I am impressed. Very well done!
